File: web/src/api/vendor-endpoint.ts
const BASE = '/api/admin'

function authHeaders(): Record<string, string> {
  const token = localStorage.getItem('token')
  return token ? { Authorization: `Bearer ${token}` } : {}
}

async function request<T>(url: string, options?: RequestInit): Promise<T> {
  const res = await fetch(url, {
    headers: { 'Content-Type': 'application/json', ...authHeaders() },
    ...options,
  })
  return res.json()
}

// ── 供应商端点 ──

export interface VendorEndpoint {
  id: number
  vendor_id: number
  endpoint_id: number
  path: string
  name: string
  description: string
  is_async: boolean
  status: number
  created_at: string
  vendor_name: string
  endpoint_path: string
  endpoint_name: string
}

export function getVendorEndpoints() {
  return request<{ success: boolean; data: VendorEndpoint[] }>(`${BASE}/vendor-endpoints`)
}

export function createVendorEndpoint(data: {
  vendor_id: number
  endpoint_id: number
  path?: string
  name?: string
  description?: string
  is_async?: boolean
}) {
  return request<{ success: boolean; message: string }>(`${BASE}/vendor-endpoints`, {
    method: 'POST',
    body: JSON.stringify(data),
  })
}

export function updateVendorEndpoint(id: number, data: {
  vendor_id?: number
  endpoint_id?: number
  path?: string
  name?: string
  description?: string
  is_async?: boolean
}) {
  return request<{ success: boolean; message: string }>(`${BASE}/vendor-endpoints/${id}`, {
    method: 'PUT',
    body: JSON.stringify(data),
  })
}

export function updateVendorEndpointStatus(id: number, status: number) {
  return request<{ success: boolean; message: string }>(`${BASE}/vendor-endpoints/${id}/status`, {
    method: 'PUT',
    body: JSON.stringify({ status }),
  })
}

export function deleteVendorEndpoint(id: number) {
  return request<{ success: boolean; message: string }>(`${BASE}/vendor-endpoints/${id}`, {
    method: 'DELETE',
  })
}

// ── 供应商端点字段 ──

export interface VendorEndpointField {
  id: number
  vendor_endpoint_id: number
  endpoint_field_id: number | null
  field_key: string
  field_name: string
  field_type: string
  required: boolean
  description: string
  sort_order: number
}

export function getVendorEndpointFields(veId: number) {
  return request<{ success: boolean; data: VendorEndpointField[] }>(`${BASE}/vendor-endpoints/${veId}/fields`)
}

export function syncVEFieldsFromEndpoint(veId: number) {
  return request<{ success: boolean; message: string }>(`${BASE}/vendor-endpoints/${veId}/fields/sync`, {
    method: 'POST',
  })
}

export function createVEField(veId: number, data: {
  field_key: string
  field_name: string
  field_type: string
  required: boolean
  description?: string
  endpoint_field_id?: number | null
}) {
  return request<{ success: boolean; message: string; data?: VendorEndpointField }>(`${BASE}/vendor-endpoints/${veId}/fields`, {
    method: 'POST',
    body: JSON.stringify(data),
  })
}

export function updateVEField(veId: number, fieldId: number, data: {
  field_key?: string
  field_name?: string
  field_type?: string
  required?: boolean
  description?: string
  endpoint_field_id?: number | null
}) {
  return request<{ success: boolean; message: string }>(`${BASE}/vendor-endpoints/${veId}/fields/${fieldId}`, {
    method: 'PUT',
    body: JSON.stringify(data),
  })
}

export function deleteVEField(veId: number, fieldId: number) {
  return request<{ success: boolean; message: string }>(`${BASE}/vendor-endpoints/${veId}/fields/${fieldId}`, {
    method: 'DELETE',
  })
}

File: model/vendor_endpoint.go
package model

import (
	"time"

	"gorm.io/gorm"
)

// VendorEndpoint 供应商端点（供应商 + 端点 + 供应商侧自定义配置）
type VendorEndpoint struct {
	ID          int        `json:"id" gorm:"primaryKey"`
	VendorID    int        `json:"vendor_id" gorm:"index;not null"`
	EndpointID  int        `json:"endpoint_id" gorm:"index;not null"`
	Path        string     `json:"path" gorm:"size:512"`         // 供应商侧路径，为空则用端点默认路径
	Name        string     `json:"name" gorm:"size:128"`         // 供应商侧名称，为空则用端点默认名称
	Description string     `json:"description" gorm:"size:512"`  // 供应商侧描述
	IsAsync     bool       `json:"is_async" gorm:"default:false"`
	Status      int        `json:"status" gorm:"default:1"`
	CreatedAt   time.Time  `json:"created_at"`
	UpdatedAt   time.Time  `json:"updated_at"`
	DeletedAt   gorm.DeletedAt `json:"-" gorm:"index"`

	// 非数据库字段
	VendorName    string `json:"vendor_name" gorm:"-"`
	EndpointPath  string `json:"endpoint_path" gorm:"-"`
	EndpointName  string `json:"endpoint_name" gorm:"-"`
}

// AdminGetVendorEndpoints 获取全部供应商端点
func AdminGetVendorEndpoints() ([]VendorEndpoint, error) {
	var items []VendorEndpoint
	err := DB.Order("id ASC").Find(&items).Error
	if err != nil {
		return nil, err
	}

	vendorMap, _ := GetVendorMapByID()
	epMap, _ := GetEndpointMapByID()

	for i := range items {
		if v, ok := vendorMap[items[i].VendorID]; ok {
			items[i].VendorName = v.Name
		}
		if ep, ok := epMap[items[i].EndpointID]; ok {
			items[i].EndpointPath = ep.Path
			items[i].EndpointName = ep.Name
		}
	}
	return items, nil
}

// GetVendorEndpointByID 根据 ID 获取供应商端点
func GetVendorEndpointByID(id int) (*VendorEndpoint, error) {
	var ve VendorEndpoint
	err := DB.Where("id = ?", id).First(&ve).Error
	if err != nil {
		return nil, err
	}
	return &ve, nil
}

// CreateVendorEndpoint 创建供应商端点
func CreateVendorEndpoint(ve *VendorEndpoint) error {
	return DB.Create(ve).Error
}

// UpdateVendorEndpoint 更新供应商端点
func UpdateVendorEndpoint(id int, updates map[string]interface{}) error {
	return DB.Model(&VendorEndpoint{}).Where("id = ?", id).Updates(updates).Error
}

// UpdateVendorEndpointStatus 更新状态
func UpdateVendorEndpointStatus(id int, status int) error {
	return DB.Model(&VendorEndpoint{}).Where("id = ?", id).Update("status", status).Error
}

// DeleteVendorEndpoint 删除
func DeleteVendorEndpoint(id int) error {
	return DB.Delete(&VendorEndpoint{}, id).Error
}
